model/dns_zone/dns_zone.rb

```diff
@@ -0,0 +1,86 @@
+# frozen_string_literal: true
+
+require "zlib"
+require_relative "../../model"
+
+class DnsZone < Sequel::Model
+  many_to_one :project
+  many_to_many :name_servers
+  one_to_many :resource_records, class: :DnsZoneResourceRecord
+  one_to_many :active_billing_records, class: :BillingRecord, key: :resource_id, conditions: {Sequel.function(:upper, :span) => nil}
+
+  include ResourceMethods
+  include Authorization::HyperTagMethods
+  include Authorization::TaggableMethods
+
+  def hyper_tag_name(project)
+    "project/#{project.ubid}/location/#{location}/dns_zone/#{name}"
+  end
+
+  def insert_record(record_name:, type:, ttl:, data:)
+    begin
+      DnsZoneResourceRecord.create_with_id(dns_zone_id: id, name: record_name, type: type, ttl: ttl, data: data)
+    rescue Sequel::ValidationFailed => ex
+      raise unless ex.message.include?("is already taken")
+    end
+
+    command = "zone-set #{name} #{record_name} #{ttl} #{type} #{data}"
+    run_on_all_name_servers(command, "such record already exists in zone")
+  end
+
+  def delete_record(record_name:, type: nil, data: nil)
+    fail "Type needs to be specified if data is specified!" if data && type.nil?
+
+    records = resource_records_dataset.where(name: record_name)
+    records = records.where(type: type) if type
+    records = records.where(data: data) if data
+    records.each do |r|
+      r.destroy
+    end
+
+    command = "zone-unset #{name} #{record_name} #{type} #{data}"
+    run_on_all_name_servers(command, "no such node in zone found")
+  end
+
+  def run_on_all_name_servers(command, ignore)
+    lock_name_servers do
+      name_servers.flat_map(&:vms).each do |vm|
+        # knotc also supports passing multiple commands via stdin, but we
+        # don't use that, because exception handling would be much more
+        # difficult with that.
+        begin
+          vm.sshable.cmd("sudo -u knot knotc zone-abort #{name}")
+        rescue Sshable::SshError => ex
+          raise unless ex.stdout.include?("no active transaction")
+        end
+
+        vm.sshable.cmd("sudo -u knot knotc zone-begin #{name}")
+
+        begin
+          vm.sshable.cmd("sudo -u knot knotc #{command}")
+        rescue Sshable::SshError => ex
+          raise unless ex.stdout.include?(ignore)
+        end
+
+        vm.sshable.cmd("sudo -u knot knotc zone-commit #{name}")
```

Comment:
How is the failure on a single vm command handled here? Is it safe to perform this on a loop? If we're not able to connect to 1 of the VMs and because of the array ordering that's let's say the first or second one, does this cause an issue? Should we find a way to do this in an asynchronous way?

Reply:
Good point, this is something I spend few different approaches during development.

We rely on retry mechanism here. If the command fails at one of the VMs for some reason, the exception would propagated to upper layers and we would retry the operation in next iteration. At that time, we would do the same modification on all VMs (even if it is already done at the first time). Initially data will be inconsistent, but after the retry is completed, data in different VMs will be consistent again.

During development, I tried 2 workarounds for this problem but eventually decided to let the retry handle it automatically. Here is what I did;
- I moved commit to a separate loop, so that we would only try to commit after ensuring other commands succeeded in all VMs. This is like poor man's 2PC commit. Of course the command can still fail during commit, but it lowers the probability.
- I created a page if the scenario you described happens, but then realized there is not much an operator can do other than retrying manually.